### PR TITLE
New algorithm to slice event workspaces for ISIS reflectometry

### PR DESCRIPTION
**Pull request by gemmaguest**
**Description of work.**
This PR adds a new algorithm to perform slicing of event workspaces, producing output suitable for use with ReflectometryReductionOneAuto.
- Uses GenerateEventsFilter and FilterEvents to do the filtering
- Rebins to histogram data
- Adds monitors to each resulting slice, scaled by the relative proton_charge for the slice

The aim of this algorithm will eventually be to replace functionality currently implemented in the ISIS Reflectometry GUI, which should simplify the restructuring work in #23051.

**Report to:** nobody

**To test:**
Try out the algorithm with various different inputs for filtering, e.g. the following should result in a group workspace with 3 slices. Check that each contains 3 monitors with scaled Y values compared to the input monitor workspace.
```
Load('OFFSPEC48276', OutputWorkspace='OFFSPEC48276',LoadMonitors=True)
ReflectometrySliceEventWorkspace(InputWorkspace='OFFSPEC48276', MonitorWorkspace='OFFSPEC48276_monitors',OutputWorkspace='TOF_48276',
    TimeInterval=800)
```

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by Mantid-Matthew** [comment]
Code and tests look good. 

I have one minor issue to do with C++ types which I'm not even sure is an error. When I run the following script
```
Load('OFFSPEC48276', OutputWorkspace='OFFSPEC48276',LoadMonitors=True)
ReflectometrySliceEventWorkspace(InputWorkspace='OFFSPEC48276', MonitorWorkspace='OFFSPEC48276_monitors',OutputWorkspace='TOF_48276',
    TimeInterval=400, StartTime=5, StopTime=500)
```

I get the following error 

```
TypeError: No registered converter was able to produce a C++ rvalue of type std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > from this Python object of type int
  at line 3 in 'New script'
  caused by line 1094 in '/home/matthew/Documents/Mantid/mantidBuild/bin/mantid/simpleapi.py'
  caused by line 1032 in '/home/matthew/Documents/Mantid/mantidBuild/bin/mantid/simpleapi.py'
  caused by line 1016 in '/home/matthew/Documents/Mantid/mantidBuild/bin/mantid/simpleapi.py'
``` 

This seems to be because GenerateEventsFilter expects a string for StartTime and StopTime passing these numbers as strings works but I was wondering if it would be better to sanatise this input into a string inside the python algorithm?

**Comment by gemmaguest** [comment]
I think I'd rather leave the inputs the same as GenerateEventsFilter for now, although I may change this in the future depending on user feedback.

**Review by Matthew-Andrew — APPROVED** [review]
OK looks good to go to me then.

File: docs/source/algorithms/ReflectometrySliceEventWorkspace-v1.rst [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    AddTimeSeriesLog(input_ws, Name="proton_charge", Time="2010-01-01T00:50:00", Value=100)
+    CreateSampleWorkspace(OutputWorkspace='monitor_ws', NumBanks=0, NumMonitors=3, BankPixelWidth=1, NumEvents=10000, Random=True)
+
+    output = ReflectometrySliceEventWorkspace(InputWorkspace='input_ws', MonitorWorkspace='monitor_ws',OutputWorkspace='sliced',
+    TimeInterval=600, StartTime='1800', StopTime='3300')
+
+    print output.getNumberOfEntries(),'slices'
+    print output[0].getNumberHistograms(),'spectra'
+
+.. seealso :: Algorithm :ref:`algm-GenerateEventsFilter`, :ref:`algm-FilterEvents` and :ref:`algm-ReflectometryReductionOneAuto`
```
Comment (AndreiSavici) [inline docs/source/algorithms/ReflectometrySliceEventWorkspace-v1.rst]: There is a function in algorithm that should be overwritten instead. Add the following to the algorithm code
```
def seeAlso(self):
    return [ "GenerateEventsFilter","FilterEvents" ,"ReflectometryReductionOneAuto"]
```

File: docs/source/algorithms/ReflectometrySliceEventWorkspace-v1.rst [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ReflectometrySliceEventWorkspace
+--------------------------------
+
+This algorithm slices an input workspace into one or more grouped output workspaces based on input filtering properties.
+
+It uses :ref:`algm-GenerateEventsFilter` to define the way splitting should be done and exposes the relevant input properties for that algorithm. It then performs the filtering using :ref:`algm-FilterEvents`.
+
+The sliced workspaces are then rebinned to histogram data and combined with the given monitor workspace, to produce a workspace suitable for input to :ref:`algm-ReflectometryReductionOneAuto`. The monitors for each slice are scaled according to the percentage of ``proton_charge`` in that slice.
+
+Usage
```
Comment (AndreiSavici) [inline docs/source/algorithms/ReflectometrySliceEventWorkspace-v1.rst]: Add a correct (testable) documentation usage.

**Review by AndreiSavici — CHANGES_REQUESTED** [review]

**Review by AndreiSavici — APPROVED** [review]

**Comment by AndreiSavici** [comment]
Unrelated systemtest error

**Pull request merged**